library/Icingadb/Util/OptimizerHints.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace Icinga\Module\Icingadb\Util;
+
+use Icinga\Application\Config;
+use ipl\Orm\Query;
+use ipl\Sql\Adapter\Mysql;
+use ipl\Sql\Select;
+use ipl\Stdlib\Str;
+
+/**
+ * Helper class to allow disabling the MySQL/MariaDB query optimizer for history queries.
+ * Some versions of these RDBMS perform poorly with history queries,
+ * particularly when the optimizer changes join order or uses block nested loop joins.
+ * Ideally, testing across all RDBMS versions to identify when the optimizer fails and adjusting queries or
+ * using optimizer switches would be preferable, but this level of effort is not justified at the moment.
+ */
+readonly class OptimizerHints
+{
+    public const DISABLE_OPTIMIZER_HINT = '/*+ NO_BNL() */ STRAIGHT_JOIN';
+
+    /**
+     * Determines whether to disable the query optimizer for history queries.
+     *
+     * Call this method before using {@see disableOptimizer()} to check if optimizer disabling is necessary:
+     *
+     * ```
+     * if (OptimizerHints::shouldDisableOptimizerForHistoryQueries()) {
+     *     OptimizerHints::disableOptimizer($query);
+     * }
+     * ```
+     *
+     * @return bool
+     */
+    public static function shouldDisableOptimizerForHistoryQueries(): bool
+    {
+        return Config::module('icingadb')->get('icingadb', 'disable_optimizer_for_history_queries', false);
+    }
+
+    /**
+     * Injects an optimizer hint into SELECT queries for a MySQL/MariaDB Query object,
+     * forcing the RDBMS to disable the optimizer.
+     *
+     * Call {@see shouldDisableOptimizerForHistoryQueries()} first to check if this is necessary:
+     *
+     *  ```
+     *  if (OptimizerHints::shouldDisableOptimizerForHistoryQueries()) {
+     *      OptimizerHints::disableOptimizer($query);
+     *  }
+     *  ```
+     *
+     * @param Query $q
+     *
+     * @return void
+     */
+    public static function disableOptimizer(Query $q): void
+    {
+        if ($q->getDb()->getAdapter() instanceof Mysql) {
+            // Locates the first string column, prepends the optimizer hint,
+            // and resets columns to ensure it appears first in the SELECT statement:
```

Comment:
why the *first* string column? Does the query hint not work if placed before an (aliased) expression?

Reply:
Sure it works, but it will also increase complexity. My thought was that the hint should work (or rather not add crap) for every query scenario, even if it can't be added at all as with `SELECT 1` (unaliased expression). Considering that all other queries will select real columns, I decided to use this implementation.